Microsoft.Toolkit.HighPerformance/Streams/MemoryStream.ThrowExceptions.cs

```diff
@@ -0,0 +1,110 @@
+            throw new ArgumentOutOfRangeException("offset", "Offset can't be negative.");
+        }
+
+        /// <summary>
+        /// Throws an <see cref="ArgumentOutOfRangeException"/> when the input count is negative.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private static void ThrowArgumentOutOfRangeExceptionForCount()
+        {
+            throw new ArgumentOutOfRangeException("count", "Count can't be negative.");
+        }
+
+        /// <summary>
+        /// Throws an <see cref="ArgumentException"/> when the sum of offset and count exceeds the length of the target buffer.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private static void ThrowArgumentExceptionForLength()
+        {
+            throw new ArgumentException("The sum of offset and count can't be larger than the buffer length.", "buffer");
+        }
+
+        /// <summary>
+        /// Throws a <see cref="NotSupportedException"/> when trying to write on a readonly stream.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private static void ThrowNotSupportedExceptionForCanWrite()
+        {
+            throw new NotSupportedException("The current stream doesn't support writing.");
+        }
+
+        /// <summary>
+        /// Throws an <see cref="InvalidOperationException"/> when trying to write too many bytes to the target stream.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private static void ThrowArgumentExceptionForEndOfStreamOnWrite()
+        {
+            throw new InvalidOperationException("The current stream can't contain the requested input data.");
+        }
+
+        /// <summary>
+        /// Throws a <see cref="NotSupportedException"/> when trying to set the length of the stream.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        public static void ThrowNotSupportedExceptionForSetLength()
+        {
+            throw new NotSupportedException("Setting the length is not supported for this stream.");
+        }
+
+        /// <summary>
+        /// Throws an <see cref="ArgumentException"/> when using an invalid seek mode.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        [SuppressMessage("StyleCop.CSharp.DocumentationRules", "SA1615", Justification = "Throw method")]
+        public static long ThrowArgumentExceptionForSeekOrigin()
+        {
+            throw new ArgumentException("The input seek mode is not valid.", "origin");
+        }
+
+        /// <summary>
+        /// Throws an <see cref="ObjectDisposedException"/> when using a disposed <see cref="Stream"/> instance.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        public static void ThrowObjectDisposedException()
+        {
+            throw new ObjectDisposedException("stream", "The current stream has already been disposed");
```

Comment:
Wait, the memory itself is not necessarily disposed though 🤔
I mean, if the stream is a `MemoryStream` wrapping a `Memory<T>`, when it's disposed it just means that the `Stream` instance itself has been disposed, but all it does in practice is just to lose the reference that was internally stored by the `Memory<T>` instance - the actual memory area is not guaranteed to be disposed. As in, consider this snippet:

```csharp
int[] array = new int[100];

var stream = array.AsMemory().AsStream();
stream.Dispose();

// The stream is disposed here, but the array is fine
```

I feel like using `"memory"` could be confusing for users: as far as they can tell it's just the stream that has been disposed, not a memory (also because `Memory<T>` is not actually disposable either). I'm open to suggestions on how to maybe phrase this better, let me know what you think! 😊

Reply:
I was more worried about someone maintaining this code, since the parameter of that exception is "objectName" and it refers to "string containing the name of the disposed object". There is no "stream" field or variable anywhere in that class that would make sense to search for, and this code path is checking for the field "memory". This method is only called on the ValidateDisposed method, which is checking for the disposal of the "memory" field. I agree that it makes more sense to be the stream, but that is called out on the message of the exception.

Reply:
Ah I see, I understand your point now, thanks! 😄
So users seeing that exception would rely on the message, and maintainers would look up the variable name if they need to investigate issues there, gotcha. Fixed in https://github.com/windows-toolkit/WindowsCommunityToolkit/pull/3128/commits/31c27e6a523348ff893b899b1e0e9615359d11f0.